File: C++/design-most-recently-used-queue.cpp
// Time:  ctor:  O(n + m), m is the max number of calls
// Space: fetch: O(log(n + m))

class MRUQueue {
public:
    MRUQueue(int n)
     : bit_{n}
     , curr_{n} {
        for (int i = 0; i < n; ++i) {
            lookup_[i] = i + 1;
        }
    }
    
    int fetch(int k) {
        int pos  = bit_.binary_lift(k);
        int val = lookup_[pos];
        lookup_.erase(pos);
        bit_.add(pos, -1);
        bit_.add(curr_, 1);
        lookup_[curr_++] = val;
        return val;
    }

private:
    class BIT {
    public:
        static const int MAX_CALLS = 2000;

        BIT(int n) : bit_(n + MAX_CALLS + 1) {  // 0-indexed
            for (int i = 1; i < size(bit_); ++i) {
                bit_[i] = ((i - 1 < n) ? 1 : 0) + bit_[i - 1];
            }
            for (int i = size(bit_) - 1; i >= 1; --i) {
                int last_i = i - lower_bit(i);
                bit_[i] -= bit_[last_i];
            }
        }
        
        void add(int i, int val) {
            ++i;
            for (; i < size(bit_); i += lower_bit(i)) {
                bit_[i] += val;
            }
        }

        int query(int i) const {
            ++i;
            int total = 0;
            for (; i > 0; i -= lower_bit(i)) {
                total += bit_[i];
            }
            return total;
        }
        
        int binary_lift(int k) const {
            int total = 0;
            int pos = 0;
            for (int i = floor_log2_x(size(bit_) - 1); i >= 0; --i) {
                if (pos + (1 << i) < size(bit_) && !(total + bit_[pos + (1 << i)] >= k)) {
                    total += bit_[pos + (1 << i)];
                    pos += (1 << i);
                }
            }
            return (pos + 1) - 1;
        }
    
    private:
        int lower_bit(int i) const {
            return i & -i;
        }
        
        int floor_log2_x(int x) const {
            return 8 * sizeof(int) - __builtin_clz(x) - 1;
        };
        
        vector<int> bit_;
    };
    
    BIT bit_;
    unordered_map<int, int> lookup_;
    int curr_;
};

// Time:  ctor:  O(n)
// Space: fetch: O(sqrt(n))
// sqrt decomposition solution
class MRUQueue2 {
public:
    MRUQueue2(int n)
     : buckets_(ceil(sqrt(n))) {
        for (int i = 0; i < n; ++i) {
            buckets_[i / size(buckets_)].emplace_back(i + 1);
        }
    }
    
    int fetch(int k) {
        --k;
        int left = k / size(buckets_);
        int idx = k % size(buckets_);
        auto cit = cbegin(buckets_[left]);
        advance(cit, idx);
        int val = *cit;
        buckets_[left].erase(cit);
        buckets_.back().emplace_back(val);
        for (int i = size(buckets_) - 2; i >= left; --i) {
            int x = buckets_[i + 1].front();
            buckets_[i + 1].pop_front();
            buckets_[i].emplace_back(x);
        }
        return val;
    }
    
private:
    vector<list<int>> buckets_;
};

File: Python/design-most-recently-used-queue.py
# Time:  ctor:  O(nlogn)
# Space: fetch: O(logn)

from sortedcontainers import SortedList


# balanced bst solution
class MRUQueue(object):

    def __init__(self, n):
        """
        :type n: int
        """
        self.__sl = SortedList((i-1, i) for i in xrange(1, n+1))  

    def fetch(self, k):
        """
        :type k: int
        :rtype: int
        """
        last, _ = self.__sl[-1]
        _, val = self.__sl.pop(k-1)
        self.__sl.add((last+1, val))
        return val


# Time:  ctor:  O(n + m), m is the max number of calls
# Space: fetch: O(log(n + m))
class BIT(object):  # 0-indexed.
    def __init__(self, n):
        MAX_CALLS = 2000
        self.__bit = [0]*(n+MAX_CALLS+1)  # Extra one for dummy node.
        for i in xrange(1, len(self.__bit)):
            self.__bit[i] = (1 if i-1 < n else 0) + self.__bit[i-1]
        for i in reversed(xrange(1, len(self.__bit))):
            last_i = i - (i & -i)
            self.__bit[i] -= self.__bit[last_i]

    def add(self, i, val):
        i += 1  # Extra one for dummy node.
        while i < len(self.__bit):
            self.__bit[i] += val
            i += (i & -i)

    def query(self, i):
        i += 1  # Extra one for dummy node.
        ret = 0
        while i > 0:
            ret += self.__bit[i]
            i -= (i & -i)
        return ret

    def binary_lift(self, k):
        floor_log2_n = (len(self.__bit)-1).bit_length()-1
        pow_i = 2**floor_log2_n
        total = pos = 0  # 1-indexed
        for i in reversed(xrange(floor_log2_n+1)):  # O(logN)
            if pos+pow_i < len(self.__bit) and not (total+self.__bit[pos+pow_i] >= k):
                total += self.__bit[pos+pow_i]
                pos += pow_i
            pow_i >>= 1
        return (pos+1)-1  # 0-indexed


# fenwick / bit solution
class MRUQueue2(object):

    def __init__(self, n):
        """
        :type n: int
        """
        self.__bit = BIT(n)
        self.__lookup = {i:i+1 for i in xrange(n)}
        self.__curr = n
        
    def fetch(self, k):
        """
        :type k: int
        :rtype: int
        """
        pos = self.__bit.binary_lift(k)  
        val = self.__lookup.pop(pos)
        self.__bit.add(pos, -1)
        self.__bit.add(self.__curr, 1)
        self.__lookup[self.__curr] = val     
        self.__curr += 1
        return val


# Time:  ctor:  O(n)
# Space: fetch: O(sqrt(n))
import collections
import math


# sqrt decomposition solution
class MRUQueue3(object):

    def __init__(self, n):
        """
        :type n: int
        """
        self.__buckets = [collections.deque() for _ in xrange(int(math.ceil(n**0.5)))]
        for i in xrange(n):
            self.__buckets[i//len(self.__buckets)].append(i+1)

    def fetch(self, k):
        """
        :type k: int
        :rtype: int
        """
        k -= 1
        left, idx = divmod(k, len(self.__buckets))
        val = self.__buckets[left][idx]
        del self.__buckets[left][idx]
        self.__buckets[-1].append(val)
        for i in reversed(xrange(left, len(self.__buckets)-1)):
            x = self.__buckets[i+1].popleft()
            self.__buckets[i].append(x)
        return val
